lib/protos.js

```diff
@@ -113,20 +119,38 @@ exports.map = function(events, event){
   var normalizedEvent = normalize(event);
 
   return foldl(function(matchingEvents, val, key, events) {
-    // If true, this is a `mixed` value, which is structured like so:
+    // There are three types of fields in metadata that can be mapped:
+    // 1) 'array' — array of strings
+    // 3) 'mixed' — array of objects
+    // 2) 'map' - object
+
+    // Mixed values are structured like so:
     //     { key: 'testEvent', value: { event: 'testEvent', someValue: 'xyz' } }
-    // We need to extract the key, which we use to match against
-    // `normalizedEvent`, and return `value` as part of `matchingEvents` if that
-    // match succeds.
+    // In their case, we need to match the normalized `key` field against
+    // `normalizedEvent`, and return whatever is in the `value` field (likely an object) as part
+    // of `matchingEvents` if that match succeds.
+
     if (type(events) === 'array') {
-      // If there's no key attached to this event mapping (unusual), skip this
-      // item.
-      if (!val.key) return matchingEvents;
-      // Extract the key and value from the `mixed` object.
-      key = val.key;
-      val = val.value;
+      switch (type(val)) {
+        // 'array' map
+        case 'string':
+          if (normalize(val) === normalizedEvent) {
+            matchingEvents.push(val);
+          }
+          break;
+        // 'mixed' map
+        case 'object':
+          if (normalize(val.key) === normalizedEvent) {
+            matchingEvents.push(val.value);
+          }
+          break;
+        default:
+          // no default case
+      }
+      return matchingEvents;
```

Comment:
This logic is pretty hairy and doesn't handle all the cases it used to, and will fail in some subtle cases. Here's a refactor I haven't run you might try out:

```js
var isMap = function(key, val) {
  return typeof val === 'string';
};

var isMixed = function(key, val) {
  return isObject(val) && has(val, 'key') && has(val, 'value');
};

var getEventMappingType = function(events) {
  if (all(events, isString) return 'array';
  if (all(events, isMap)) return 'map';
  if (all(events, isMixed) return 'mixed';
  return 'unknown';
};

Integration.prototype.map = function(events, event) {
  var mappingType = getEventMappingType(events);
  var normalizedEvent = normalize(event);

  if (mappingType === 'unknown') {
    return [];
  }

  return foldl(function(matchingEvents, val, key) {
    var compare;
    var result;

    if (mappingType === 'array') {
      compare = val;
      result = val;
    }
    if (mappingType === 'map') {
      compare = key;
      result = val;
    }
    if (mappingType === 'mixed') {
      compare = val.key;
      result = val.value;
    }

    if (normalize(compare) === normalizedEvent) {
      matchingEvents.push(result);
    }
    return matchingEvents;
  }, [], events);
};

```

This would all be _way_ simpler if metadata were sent down as `type: "map|array|mixed", data: ...` rather than us having to punch this till it works—we should investigate that when we look into the `mixed` refactor.